From 64d4d8e6f99bb5e564129ec5d156de8e76edde65 Mon Sep 17 00:00:00 2001
From: mpaluri <mpaluri@qti.qualcomm.com>
Date: Wed, 2 Sep 2026 14:17:56 -0700
Subject: [PATCH] db: Add support to store multiple descriptors per client in
 EndToEndQMDescriptor datamodel

Add support to add multiple descriptors each uniquely identified by QMID
and Client MAC in EndotEndQMDescriptor datamodel. Also clear the
descriptors based on QMID and Client MAC from the data model.

Signed-off-by: Manaswini Paluri <mpaluri@qti.qualcomm.com>

diff --git a/controller/src/beerocks/master/db/db.cpp b/controller/src/beerocks/master/db/db.cpp
index 97c02eb7e..692b17b3a 100644
--- a/controller/src/beerocks/master/db/db.cpp
+++ b/controller/src/beerocks/master/db/db.cpp
@@ -8337,8 +8337,12 @@ bool db::set_endtoend_qm_descriptor(const sMacAddr &client_mac, uint16_t qmid,
     const std::string e2e_descriptor_path =
         std::string(DATAELEMENTS_ROOT_DM) + ".Network.EndToEndQMDescriptor";
 
-    auto index = m_ambiorix_datamodel->get_instance_index(
-        e2e_descriptor_path + ".[ClientMAC == '%s'].", tlvf::mac_to_string(client_mac));
+    /* Look up by composite key (ClientMAC, QMID) so multiple descriptors
+     * for the same client coexist as separate DM instances. */
+    const auto search_path = e2e_descriptor_path +
+        ".[ClientMAC == '" + tlvf::mac_to_string(client_mac) +
+        "' && QMID == " + std::to_string(static_cast<unsigned>(qmid)) + "].";
+    auto index = m_ambiorix_datamodel->get_instance_index(search_path, "");
 
     std::string instance_path;
     if (index) {
@@ -8361,12 +8365,14 @@ bool db::set_endtoend_qm_descriptor(const sMacAddr &client_mac, uint16_t qmid,
     return ret_val;
 }
 
-bool db::remove_endtoend_qm_descriptor(const sMacAddr &client_mac)
+bool db::remove_endtoend_qm_descriptor(const sMacAddr &client_mac, uint16_t qmid)
 {
     const std::string e2e_descriptor_path =
         std::string(DATAELEMENTS_ROOT_DM) + ".Network.EndToEndQMDescriptor";
-    auto index = m_ambiorix_datamodel->get_instance_index(
-        e2e_descriptor_path + ".[ClientMAC == '%s'].", tlvf::mac_to_string(client_mac));
+    const auto search_path = e2e_descriptor_path +
+        ".[ClientMAC == '" + tlvf::mac_to_string(client_mac) +
+        "' && QMID == " + std::to_string(static_cast<unsigned>(qmid)) + "].";
+    auto index = m_ambiorix_datamodel->get_instance_index(search_path, "");
     if (!index) {
         return true;
     }
@@ -8493,8 +8499,8 @@ bool db::clear_transient_controller_qm_descriptors(const sMacAddr &al_mac)
 
         for (auto client_entry = bss_entry->second.begin();
              client_entry != bss_entry->second.end();) {
-            bool any_bss_level_removed     = false;
-            bool any_network_level_removed = false;
+            bool any_bss_level_removed = false;
+            std::vector<uint16_t> network_qmids_to_remove;
             for (auto key_entry = client_entry->second.begin();
                  key_entry != client_entry->second.end();) {
                 if (key_entry->second.request_type !=
@@ -8504,7 +8510,7 @@ bool db::clear_transient_controller_qm_descriptors(const sMacAddr &al_mac)
                 }
                 if (key_entry->second.descriptor_config_type ==
                     eDescriptorConfigType::NETWORK_LEVEL) {
-                    any_network_level_removed = true;
+                    network_qmids_to_remove.push_back(key_entry->second.qmid);
                 } else {
                     any_bss_level_removed = true;
                 }
@@ -8518,10 +8524,10 @@ bool db::clear_transient_controller_qm_descriptors(const sMacAddr &al_mac)
                     ret_val = false;
                 }
             }
-            if (any_network_level_removed) {
-                if (!remove_endtoend_qm_descriptor(client_entry->first)) {
+            for (const uint16_t qmid : network_qmids_to_remove) {
+                if (!remove_endtoend_qm_descriptor(client_entry->first, qmid)) {
                     LOG(ERROR) << "Failed removing transient E2E QMDescriptor for client "
-                               << client_entry->first;
+                               << client_entry->first << " qmid=" << qmid;
                     ret_val = false;
                 }
             }
diff --git a/controller/src/beerocks/master/db/db.h b/controller/src/beerocks/master/db/db.h
index f1d747a24..854326955 100644
--- a/controller/src/beerocks/master/db/db.h
+++ b/controller/src/beerocks/master/db/db.h
@@ -2539,10 +2539,11 @@ public:
     /**
      * @brief Remove a network-level (end-to-end) QoS management descriptor from the Data Model.
      *
-     * @param[in] client_mac MAC address of the end client for which the descriptor applies.
+     * @param[in] client_mac MAC address of the end client.
+     * @param[in] qmid Controller-assigned QoS Management ID of the descriptor to remove.
      * @return True on success, otherwise false.
      */
-    bool remove_endtoend_qm_descriptor(const sMacAddr &client_mac);
+    bool remove_endtoend_qm_descriptor(const sMacAddr &client_mac, uint16_t qmid);
 
     /**
      * @brief Remove a QoS management descriptor from the Data Model.
-- 
2.34.1

